source/slang/slang-emit-cpp.cpp

```diff
@@ -1561,7 +1571,117 @@ void CPPSourceEmitter::emitParamTypeImpl(IRType* type, String const& name)
 
 void CPPSourceEmitter::emitWitnessTable(IRWitnessTable* witnessTable)
 {
-    SLANG_UNUSED(witnessTable);
+    auto interfaceType = cast<IRInterfaceType>(witnessTable->getOperand(0));
+    auto witnessTableItems = witnessTable->getChildren();
+    List<IRWitnessTableEntry*> sortedWitnessTableEntries = getSortedWitnessTableEntries(witnessTable);
+    _maybeEmitWitnessTableTypeDefinition(interfaceType, sortedWitnessTableEntries);
+
+    // Define a global variable for the witness table.
+    m_writer->emit("extern ");
+    emitSimpleType(interfaceType);
+    m_writer->emit(" ");
+    m_writer->emit(getName(witnessTable));
+    m_writer->emit(";\n");
+
+    // The actual definition of this witness table global variable
+    // is deferred until the entire `Context` class is emitted, so
+    // that the member functions are available for reference.
+    // The witness table definition emission logic is defined in the
+    // `_emitWitnessTableDefinitions` function.
+    pendingWitnessTableDefinitions.add(witnessTable);
+}
+
+void CPPSourceEmitter::_emitWitnessTableDefinitions()
+{
+    for (auto witnessTable : pendingWitnessTableDefinitions)
+    {
+        auto interfaceType = cast<IRInterfaceType>(witnessTable->getOperand(0));
+        List<IRWitnessTableEntry*> sortedWitnessTableEntries = getSortedWitnessTableEntries(witnessTable);
+        emitSimpleType(interfaceType);
+        m_writer->emit(" ");
+        m_writer->emit(getName(witnessTable));
+        m_writer->emit(" = {\n");
+        m_writer->indent();
+        bool isFirstEntry = true;
+        for (Index i = 0; i < sortedWitnessTableEntries.getCount(); i++)
+        {
+            auto entry = sortedWitnessTableEntries[i];
+            if (auto funcVal = as<IRFunc>(entry->satisfyingVal.get()))
+            {
+                if (!isFirstEntry)
+                    m_writer->emit(",\n");
+                else
+                    isFirstEntry = false;
+                m_writer->emit("&Context::");
+                m_writer->emit(getName(funcVal));
+            }
+            else
+            {
+                // TODO: handle other witness table entry types.
+            }
+        }
+        m_writer->dedent();
+        m_writer->emit("\n};\n");
+    }
+}
+
+void CPPSourceEmitter::emitInterface(IRInterfaceType* interfaceType)
+{
+    // The current IRInterfaceType defintion does not contain
+    // sufficient info for emitting a witness table struct by itself
+    // Instead, it defines the order of entries in a witness table.
+    // Therefore, we emit a forward declaration here, and actual definition
+    // for the witness table type during emitWitnessTable.
```

Comment:
This seems a bit worrying in the long run, since it amounts to saying that the first witness table for an interface gets to dictate the signature of the interface methods for all other implementations of that interface.

It seems like we will need to work towards a situation where the interface type records not only the keys for its requirements, but also the expected type of the value that satisfies the requirement. The big challenge there is what type should be used to stand in for the `This` type, associated types, etc.

This is obviously something we should defer solving until it becomes critical.